README.md
# logstash-kafka

### Status
[![Build Status](https://travis-ci.org/joekiller/logstash-kafka.svg?branch=master)](https://travis-ci.org/joekiller/logstash-kafka)

This project implements Kafka 0.8.2.1 inputs and outputs for logstash 1.4.X only. For 1.5 support read below.

For more info about logstash, see <http://logstash.net/>

logstash-kafka has been intergrated into [logstash-input-kafka](https://github.com/logstash-plugins/logstash-input-kafka) and [logstash-output-kafka](https://github.com/logstash-plugins/logstash-output-kafka).  
It will be released with the 1.5 version of logstash.  Thank you for your support.  My goal is to close out tickets here 
but for the most part, issues and problems should be resolved via:

  * https://groups.google.com/forum/#!forum/logstash-users The logstash-users mailing list is extremely active and I 
  participate especially in Kafka troubleshooting/discussion there.
  
  * IRC: #logstash on irc.freenode.org
  
  * If you really have an issue that you can't resolve via the mailing list or IRC channel, try posting an issue to
  github.  Please make sure that issues are actually software issues, not configuration issues as those should be left
  to the mailing lists and IRC. https://github.com/elasticsearch/logstash/issues

I will keep this library up to date at least until logstash 1.5 is released.

I will continue helping with the plugin and readying for the next version of Kafka with the jruby-kafka library.

## Dependencies

* [Apache Kafka] version 0.8.2.1 

* [jruby-kafka] library.

[Apache Kafka]: http://kafka.apache.org/
[jruby-kafka]: https://github.com/joekiller/jruby-kafka

## Building Logstash Tarball

Because this is a plugin to Logstash, it must be built.  Luckily for you, there is a make file that handles all of this.

Most of the logic originated from logstash's make file so thank you everyone who had contributed to it to enable me to
make this easy for you.

To simply build the logstash jar as is with Kafka enabled run:

    # make tarball

## Building Logstash-Kafka RPM

**Note that this doesn't build a logstash RPM but an RPM that will install the logstash-kafka libraries on top of an existing logstash installation**

To build an rpm

    # make package

Installing the resulting rpm after installing logstash from the elasticsearch repo will copy the kafka plugin and dependencies into `/opt/logstash`.


## Building Logstash-Kafka Gem

You can build the gem file and then install it to logstash.

    # gem build logstash-kafka.gemspec

## Installing Gem to Logstash

You can install your own build logstash-kafka gem or use the one published to Rubygems. This is the easiest way to manually install the plugin regardless of if you are using Windows, Linux, BSD, etc.

### Rubygems Logstash-Kafka install for Logstash 1.5+

Please see [logstash-input-kafka] and [logstash-output-kafka] for Logstash 1.5+

### Rubygems Logstash-Kafka install on Logstash 1.4.2

```bash
cd /path/to/logstash
GEM_HOME=vendor/bundle/jruby/1.9 GEM_PATH= java -jar vendor/jar/jruby-complete-1.7.11.jar -S gem install logstash-kafka
cp -R vendor/bundle/jruby/1.9/gems/logstash-kafka-*-java/{lib/logstash/*,spec/*} {lib/logstash/,spec/}
# test install
USE_JRUBY=1 bin/logstash rspec spec/**/kafka*.rb
```
 

### Manual Logstash-Kafka Gem Install on Logstash 1.4.2

As root or using sudo:

```bash
cd /path/to/logstash
GEM_HOME=vendor/bundle/jruby/1.9 GEM_PATH= java -jar vendor/jar/jruby-complete-1.7.11.jar -S gem install logstash-kafka-0.7.0-java.gem
cp -R vendor/bundle/jruby/1.9/gems/logstash-kafka-*-java/{lib/logstash/*,spec/*} {lib/logstash/,spec/}
```

## Configuration for runtime

jruby-kafka supports nearly all the configuration options of a Kafka high level consumer but some have been left out of
this plugin simply because either it was a priority or I hadn't tested it yet.  If one isn't currently, it should be
trivial to add it via jruby-kafka and then in the logstash input or output.

## Running

You should run this version of logstash the same as you would the normal logstash with:

    # bin/logstash -f logstash.conf

Contributed plugins can also still be installed for Logstash 1.4+ using:

    # bin/plugin install contrib

### Input

See http://kafka.apache.org/documentation.html#consumerconfigs for details about the Kafka consumer options.

    # input {
        kafka {
            zk_connect => ... # string (optional), default: "localhost:2181"
            group_id => ... # string (optional), default: "logstash"
            topic_id => ... # string (optional), default: "test"
            reset_beginning => ... # boolean (optional), default: false
            consumer_threads => ... # number (optional), default: 1
            queue_size => ... # number (optional), default: 20
            rebalance_max_retries => ... # number (optional), default: 4
            rebalance_backoff_ms => ... # number (optional), default:  2000
            consumer_timeout_ms => ... # number (optional), default: -1
            consumer_restart_on_error => ... # boolean (optional), default: true
            consumer_restart_sleep_ms => ... # number (optional), default: 0
            decorate_events => ... # boolean (optional), default: false
            consumer_id => ... # string (optional) default: nil
            fetch_message_max_bytes => ... # number (optional) default: 1048576
        }
    }

### Output

See http://kafka.apache.org/documentation.html#producerconfigs for details about the Kafka producer options.

    # output {
        kafka {
            broker_list => ... # string (optional), default: "localhost:9092"
            topic_id => ... # string (optional), default: "test"
            compression_codec => ... # string (optional), one of ["none", "gzip", "snappy"], default: "none"
            compressed_topics => ... # string (optional), default: ""
            request_required_acks => ... # number (optional), one of [-1, 0, 1], default: 0
            serializer_class => ... # string, (optional) default: "kafka.serializer.StringEncoder"
            partitioner_class => ... # string (optional) default: "kafka.producer.DefaultPartitioner"
            request_timeout_ms => ... # number (optional) default: 10000
            producer_type => ... # string (optional), one of ["sync", "async"] default => 'sync'
            key_serializer_class => ... # string (optional) default: "kafka.serializer.StringEncoder"
            message_send_max_retries => ... # number (optional) default: 3
            retry_backoff_ms => ... # number (optional) default: 100
            topic_metadata_refresh_interval_ms => ... # number (optional) default: 600 * 1000
            queue_buffering_max_ms => ... # number (optional) default: 5000
            queue_buffering_max_messages => ... # number (optional) default: 10000
            queue_enqueue_timeout_ms => ... # number (optional) default: -1
            batch_num_messages => ... # number (optional) default: 200
            send_buffer_bytes => ... # number (optional) default: 100 * 1024
            client_id => ... # string (optional) default: ""
        }
    }

The default codec is json for input and outputs.  If you select a codec of plain, logstash will encode your messages with not only the message
but also with a timestamp and hostname.  If you do not want anything but your message passing through, you should make
the output configuration something like:

    # output {
        kafka {
            codec => plain {
                format => "%{message}"
            }
        }
    }

## Testing

    # make tarball
    # make tarball-test

## Notes

The make file is updated to work with Logstash 1.4.0+.  DEB package building isn't supported at this time.


CHANGELOG
0.8.3 (April 24, 2015)
    # general
    - fix overly strict binding to jruby-kafka 1.4.0
0.8.2 (April 24, 2015)
    # general
    - include test spec files in gem
0.8.1 (April 24, 2015)
    # general
    - fix accidental release with old jruby-kafka lib
0.8.0 (April 23, 2015)
    # general
    - reverted features to ensure Logstash 1.4.2 compatibility.
    # outputs
    - removed partition key formatting
    # inputs
    - removed Event#sprintf placeholders support for topic_id
0.6.0 (September 20, 2014)
    # general
    - updated to match elasticsearch/logstash current
0.5.2 (June 10, 2014)
    # general
    - updated logstash to version 1.4.2
    - added Apache 2.0 License
    # tests
    - added tests for input and output (thank you suyograo)
    - updated tests for logstash 1.4.2
0.5.1 (May 27, 2014)
    # inputs
    - added fetch_message_max_bytes
0.5.0 (April 16, 2014)
    # general
    - update README to reflect new input and output options
    # inputs
    - added consumer_id
    # outputs
    - added options:
        * request_required_acks
        * serializer_class
        * partitioner_class
        * request_timeout_ms
        * producer_type
        * key_serializer_class
        * message_send_max_retries
        * retry_backoff_ms
        * topic_metadata_refresh_interval_ms
        * queue_buffering_max_ms
        * queue_buffering_max_messages
        * queue_enqueue_timeout_ms
        * batch_num_messages
        * send_buffer_bytes
        * client_id
    - see http://kafka.apache.org/documentation.html#producerconfigs for option details.
0.4.2 (April 8, 2014)
    # general
    - updated README to reflect Logstash 1.4.0 support.
    # inputs
    - fixed plugin to locate vendor logstash jars prior to importing jruby-kafka
    # outputs
    - fixed plugin to locate vendor logstash jars prior to importing jruby-kafka
0.4.1 (April 8, 2014)
    # general
    - updated to support Logstash 1.4.0
0.4.0 (March 13, 2014)
    # general
    - updated to support Kafka 0.8.1
    - made default codec JSON on inputs and outputs.
    # outputs
    - added support for request_required_acks
0.3.0 (January 23, 2014)
    # general
    - updated jruby-kafka gemspec dependency
    # outputs
    - added support for compression_codec and compressed_topics
0.2.0 (January 23, 2014)
    # inputs
    - fixed an issue where shutdown caused an extra message to be logged.
0.1.0 (January 13, 2014)
    # general
    - Adding change log.
    - First tagged verison of logstash-kafka.


spec/inputs/kafka.rb
# encoding: utf-8

require 'rspec'
require 'insist'
require 'logstash/namespace'
require 'logstash/inputs/kafka'
require 'logstash/errors'

describe LogStash::Inputs::Kafka do
  extend LogStash::RSpec

  let (:kafka_config) {{"topic_id" => "test"}}

  it 'should populate kafka config with default values' do
    kafka = LogStash::Inputs::Kafka.new(kafka_config)
    insist {kafka.zk_connect} == "localhost:2181"
    insist {kafka.topic_id} == "test"
    insist {kafka.group_id} == "logstash"
    insist {kafka.reset_beginning} == false
  end

  it "should register and load kafka jars without errors" do
    kafka = LogStash::Inputs::Kafka.new(kafka_config)
    kafka.register
  end

  it "should retrieve event from kafka" do
    # Extend class to control behavior
    class LogStash::Inputs::TestKafka < LogStash::Inputs::Kafka
      milestone 1
      private
      def queue_event(msg, output_queue)
        super(msg, output_queue)
        # need to raise exception here to stop the infinite loop
        raise LogStash::ShutdownSignal
      end
    end

    kafka = LogStash::Inputs::TestKafka.new(kafka_config)
    kafka.register

    class Kafka::Group
      public
      def run(a_numThreads, a_queue)
        a_queue << OpenStruct.new(:message => "Kafka message")
      end
    end

    logstash_queue = Queue.new
    kafka.run logstash_queue
    e = logstash_queue.pop
    insist { e["message"] } == "Kafka message"
    # no metadata by default
    insist { e["kafka"] } == nil
  end

end


spec/outputs/kafka.rb
# encoding: utf-8

require 'rspec'
require 'insist'
require 'logstash/namespace'
require 'time'
require 'logstash/outputs/kafka'

describe LogStash::Outputs::Kafka do

  it 'should populate kafka config with default values' do
    kafka = LogStash::Outputs::Kafka.new
    insist {kafka.broker_list} == "localhost:9092"
    insist {kafka.topic_id} == "test"
    insist {kafka.compression_codec} == "none"
    insist {kafka.serializer_class} == "kafka.serializer.StringEncoder"
    insist {kafka.partitioner_class} == "kafka.producer.DefaultPartitioner"
    insist {kafka.producer_type} == "sync"
  end

  it "should load register and load kafka jars without errors" do
    kafka = LogStash::Outputs::Kafka.new
    kafka.register
  end

  it "should send logstash event to kafka broker" do
    timestamp = Time.now
    expect_any_instance_of(Kafka::Producer)
    .to receive(:sendMsg)
        .with("test", nil, "{\"message\":\"hello world\",\"host\":\"test\",\"@timestamp\":\"#{timestamp.iso8601(3)}\",\"@version\":\"1\"}")
    e = LogStash::Event.new({"message" => "hello world", "host" => "test", "@timestamp" => timestamp})
    kafka = LogStash::Outputs::Kafka.new
    kafka.register
    kafka.receive(e)
  end

end


spec/inputs/kafka_spec.rb
require "test_utils"
require "logstash/inputs/kafka"
require 'jruby-kafka'

class LogStash::Inputs::TestKafka < LogStash::Inputs::Kafka
  private
  def queue_event(msg, output_queue)
    super(msg, output_queue)
    # need to raise exception here to stop the infinite loop
    raise LogStash::ShutdownSignal
  end
end

class TestMessageAndMetadata
  attr_reader :topic, :partition, :key, :message
  def initialize(topic, partition, key, message)
    @topic = topic
    @partition = partition
    @key = key
    @message = message
  end
end


class TestKafkaGroup < Kafka::Group
  def run(a_num_threads, a_queue)
    blah = TestMessageAndMetadata.new(@topic, 0, nil, 'Kafka message')
    a_queue << blah
  end
end

describe 'inputs/kafka' do
  let (:kafka_config) {{'topic_id' => 'test'}}
  let (:empty_config) {{}}
  let (:bad_kafka_config) {{'topic_id' => 'test', 'white_list' => 'other_topic'}}
  let (:white_list_kafka_config) {{'white_list' => 'other_topic'}}
  let (:decorated_kafka_config) {{'topic_id' => 'test', 'decorate_events' => true}}

  it "should register" do
    input = LogStash::Plugin.lookup("input", "kafka").new(kafka_config)
    expect {input.register}.to_not raise_error
  end

  it "should register with whitelist" do
    input = LogStash::Plugin.lookup("input", "kafka").new(white_list_kafka_config)
    expect {input.register}.to_not raise_error
  end

  it "should fail with multiple topic configs" do
    input = LogStash::Plugin.lookup("input", "kafka").new(empty_config)
    expect {input.register}.to raise_error
  end

  it "should fail without topic configs" do
    input = LogStash::Plugin.lookup("input", "kafka").new(bad_kafka_config)
    expect {input.register}.to raise_error
  end

  it 'should populate kafka config with default values' do
    kafka = LogStash::Inputs::TestKafka.new(kafka_config)
    insist {kafka.zk_connect} == 'localhost:2181'
    insist {kafka.topic_id} == 'test'
    insist {kafka.group_id} == 'logstash'
    !insist { kafka.reset_beginning }
  end

  it 'should retrieve event from kafka' do
    kafka = LogStash::Inputs::TestKafka.new(kafka_config)
    expect(kafka).to receive(:create_consumer_group) do |options|
      TestKafkaGroup.new(options)
    end
    kafka.register

    logstash_queue = Queue.new
    kafka.run logstash_queue
    e = logstash_queue.pop
    insist { e['message'] } == 'Kafka message'
    # no metadata by default
    insist { e['kafka'] } == nil
  end

  it 'should retrieve a decorated event from kafka' do
    kafka = LogStash::Inputs::TestKafka.new(decorated_kafka_config)
    expect(kafka).to receive(:create_consumer_group) do |options|
      TestKafkaGroup.new(options)
    end
    kafka.register

    logstash_queue = Queue.new
    kafka.run logstash_queue
    e = logstash_queue.pop
    insist { e['message'] } == 'Kafka message'
    # no metadata by default
    insist { e['kafka']['topic'] } == 'test'
    insist { e['kafka']['consumer_group'] } == 'logstash'
    insist { e['kafka']['msg_size'] } == 13
    insist { e['kafka']['partition'] } == 0
    insist { e['kafka']['key'] } == nil
  end

end

gembag.rb
#!/usr/bin/env ruby

# target for now
rubymajor = RUBY_VERSION.split(".")[0..1].join(".")
target = "#{Dir.pwd}/vendor/bundle/#{RUBY_ENGINE}/#{rubymajor}/"
ENV["GEM_HOME"] = target
ENV["GEM_PATH"] = ""

require "rubygems/specification"
require "rubygems/commands/install_command"

def install_gem(name, requirement, target)
  puts "Fetching and installing gem: #{name} (#{requirement})"

  installer = Gem::Commands::InstallCommand.new
  installer.options[:generate_rdoc] = false
  installer.options[:generate_ri] = false
  installer.options[:version] = requirement
  installer.options[:args] = [name]
  installer.options[:install_dir] = target
 
  # ruby 2.0.0 / rubygems 2.x; disable documentation generation
  installer.options[:document] = []
  begin
    installer.execute
  rescue Gem::SystemExitException => e
    if e.exit_code != 0
      puts "Installation of #{name} failed"
      raise
    end
  end
end # def install_gem

gemspec = ARGV.shift || "logstash-kafka.gemspec"

spec = Gem::Specification.load(gemspec)
deps = [spec.runtime_dependencies].flatten

deps.each do |dep|
  # TODO(sissel): Hack for now
  next if "#{dep}" == "addressable (~> 2.2.6)" 

  begin
    # Check if the gem is available
    # 'gem' returns 'true' if it loaded it, false if already loaded,
    # and raises a Gem::LoadError exception on failure.
    # Skip downloading/installing it if it's already here.
    gem(dep.name, dep.requirement)

    # If we get here, we have the gem.
    puts "Gem found matching: #{dep}"
  rescue Gem::LoadError => e
    # Not installed, continue.
    message = e.to_s

    # Sometimes we failed to load because gembag installs too
    # many things. Like 'shoulda' fails to load because two
    # conflicting versions of 'mocha' were installed.
    # Fundamentally, gembag should build a dependency graph and
    # resolve all version requirements to single nodes to prevent
    # this madness.
    #
    # Possible we can steal bundler's implementation of this,
    # or just use bundler to do it, but only if bundler doesn't
    # bite me in the ass again :)
    case message
      when /Unable to activate/
        puts "Gem found, but funky: #{dep} (#{e})"
      when /Could not find/
        puts "Gem not found: #{dep}"
        install_gem(dep.name, dep.requirement, target)
      else
        puts "Unexpected error: #{e}"
        exit 1
    end # case message
  end # begin / rescue Gem::LoadError
end # deps.each





pkg/build.sh
#!/bin/bash

basedir=$(dirname $0)/../
workdir=.

# Get version details
[ ! -f "$basedir/.VERSION.mk" ] && make -C $basedir .VERSION.mk
. $basedir/.VERSION.mk

if [ "$#" -ne 2 ] ; then
  echo "Usage: $0 os release"
  echo
  echo "Example: $0 ubuntu 12.10"
fi

silent() {
  "$@" > /dev/null 2>&1
}
log() {
  echo "$@" >&2
}

os=$1
osver=$2

_fpm() {
  target=$1
  fpm -s dir -n logstash-kafka \
    -a noarch --url "https://github.com/joekiller/logstash-kafka.git" \
    --description "Logstash Kafka" \
    --vendor "na" \
    --license "Apache 2.0" \
    "$@"
}

case $os in
  centos|fedora|redhat|sl)
        _fpm -s dir -t rpm -n "logstash-kafka" -a all -v 1.2.1 --prefix /opt/logstash lib/ vendor/
    ;;
  ubuntu|debian)
    if ! echo $RELEASE | grep -q '\.(dev\|rc.*)'; then
      # This is a dev or RC version... So change the upstream version
      # example: 1.2.2.dev => 1.2.2~dev
      # This ensures a clean upgrade path.
      RELEASE="$(echo $RELEASE | sed 's/\.\(dev\|rc.*\)/~\1/')"
    fi

    _fpm -t deb --deb-user root --deb-group root \
      --iteration "1-$REVISION" --deb-ignore-iteration-in-dependencies \
      -d "logstash = $RELEASE" -v "$RELEASE" \
      -f -C $workdir/tarball --prefix /opt/logstash $(cat $workdir/files)
    ;;
esac


lib/logstash/outputs/kafka.rb
require 'logstash/namespace'
require 'logstash/outputs/base'
require 'jruby-kafka'

# Write events to a Kafka topic. This uses the Kafka Producer API to write messages to a topic on
# the broker.
#
# The only required configuration is the topic name. The default codec is json,
# so events will be persisted on the broker in json format. If you select a codec of plain,
# Logstash will encode your messages with not only the message but also with a timestamp and
# hostname. If you do not want anything but your message passing through, you should make the output
# configuration something like:
#     output {
#       kafka {
#         codec => plain {
#            format => "%{message}"
#         }
#       }
#     }
# For more information see http://kafka.apache.org/documentation.html#theproducer
#
# Kafka producer configuration: http://kafka.apache.org/documentation.html#producerconfigs
class LogStash::Outputs::Kafka < LogStash::Outputs::Base
  config_name 'kafka'
  milestone 1
  
  attr_accessor :exception_repeats
  
  default :codec, 'json'
  # This is for bootstrapping and the producer will only use it for getting metadata (topics,
  # partitions and replicas). The socket connections for sending the actual data will be
  # established based on the broker information returned in the metadata. The format is
  # host1:port1,host2:port2, and the list can be a subset of brokers or a VIP pointing to a
  # subset of brokers.
  config :broker_list, :validate => :string, :default => 'localhost:9092'
  # The topic to produce the messages to
  config :topic_id, :validate => :string, :required => true
  # This parameter allows you to specify the compression codec for all data generated by this
  # producer. Valid values are "none", "gzip" and "snappy".
  config :compression_codec, :validate => %w( none gzip snappy ), :default => 'none'
  # This parameter allows you to set whether compression should be turned on for particular
  # topics. If the compression codec is anything other than NoCompressionCodec,
  # enable compression only for specified topics if any. If the list of compressed topics is
  # empty, then enable the specified compression codec for all topics. If the compression codec
  # is NoCompressionCodec, compression is disabled for all topics
  config :compressed_topics, :validate => :string, :default => ''
  # This value controls when a produce request is considered completed. Specifically,
  # how many other brokers must have committed the data to their log and acknowledged this to the
  # leader. For more info, see -- http://kafka.apache.org/documentation.html#producerconfigs
  config :request_required_acks, :validate => [-1,0,1], :default => 0
  # The serializer class for messages. The default encoder takes a byte[] and returns the same byte[]
  config :serializer_class, :validate => :string, :default => 'kafka.serializer.StringEncoder'
  # The partitioner class for partitioning messages amongst partitions in the topic. The default
  # partitioner is based on the hash of the key. If the key is null,
  # the message is sent to a random partition in the broker.
  # NOTE: topic_metadata_refresh_interval_ms controls how long the producer will distribute to a
  # partition in the topic. This defaults to 10 mins, so the producer will continue to write to a
  # single partition for 10 mins before it switches
  config :partitioner_class, :validate => :string, :default => 'kafka.producer.DefaultPartitioner'
  # The amount of time the broker will wait trying to meet the request.required.acks requirement
  # before sending back an error to the client.
  config :request_timeout_ms, :validate => :number, :default => 10000
  # This parameter specifies whether the messages are sent asynchronously in a background thread.
  # Valid values are (1) async for asynchronous send and (2) sync for synchronous send. By
  # setting the producer to async we allow batching together of requests (which is great for
  # throughput) but open the possibility of a failure of the client machine dropping unsent data.
  config :producer_type, :validate => %w( sync async ), :default => 'sync'
  # The serializer class for keys (defaults to the same as for messages if nothing is given)
  config :key_serializer_class, :validate => :string, :default => 'kafka.serializer.StringEncoder'
  # This property will cause the producer to automatically retry a failed send request. This
  # property specifies the number of retries when such failures occur. Note that setting a
  # non-zero value here can lead to duplicates in the case of network errors that cause a message
  # to be sent but the acknowledgement to be lost.
  config :message_send_max_retries, :validate => :number, :default => 3
  # Before each retry, the producer refreshes the metadata of relevant topics to see if a new
  # leader has been elected. Since leader election takes a bit of time,
  # this property specifies the amount of time that the producer waits before refreshing the
  # metadata.
  config :retry_backoff_ms, :validate => :number, :default => 100
  # The producer generally refreshes the topic metadata from brokers when there is a failure
  # (partition missing, leader not available...). It will also poll regularly (default: every
  # 10min so 600000ms). If you set this to a negative value, metadata will only get refreshed on
  # failure. If you set this to zero, the metadata will get refreshed after each message sent
  # (not recommended). Important note: the refresh happen only AFTER the message is sent,
  # so if the producer never sends a message the metadata is never refreshed
  config :topic_metadata_refresh_interval_ms, :validate => :number, :default => 600 * 1000
  # Maximum time to buffer data when using async mode. For example a setting of 100 will try to
  # batch together 100ms of messages to send at once. This will improve throughput but adds
  # message delivery latency due to the buffering.
  config :queue_buffering_max_ms, :validate => :number, :default => 5000
  # The maximum number of unsent messages that can be queued up the producer when using async
  # mode before either the producer must be blocked or data must be dropped.
  config :queue_buffering_max_messages, :validate => :number, :default => 10000
  # The amount of time to block before dropping messages when running in async mode and the
  # buffer has reached queue.buffering.max.messages. If set to 0 events will be enqueued
  # immediately or dropped if the queue is full (the producer send call will never block). If set
  # to -1 the producer will block indefinitely and never willingly drop a send.
  config :queue_enqueue_timeout_ms, :validate => :number, :default => -1
  # The number of messages to send in one batch when using async mode. The producer will wait
  # until either this number of messages are ready to send or queue.buffer.max.ms is reached.
  config :batch_num_messages, :validate => :number, :default => 200
  # Socket write buffer size
  config :send_buffer_bytes, :validate => :number, :default => 100 * 1024
  # The client id is a user-specified string sent in each request to help trace calls. It should
  # logically identify the application making the request.
  config :client_id, :validate => :string, :default => ""

  config :logstash_stop_on_exception_repeat, :validate => :number, :default=>0 # stop logstash after an exception repeats for certain times, 0 for never stop
  
  
  public
  def register
    LogStash::Logger.setup_log4j(@logger)
    options = {
      :broker_list => @broker_list,
      :compression_codec => @compression_codec,
      :compressed_topics => @compressed_topics,
      :request_required_acks => @request_required_acks,
      :serializer_class => @serializer_class,
      :partitioner_class => @partitioner_class,
      :request_timeout_ms => @request_timeout_ms,
      :producer_type => @producer_type,
      :key_serializer_class => @key_serializer_class,
      :message_send_max_retries => @message_send_max_retries,
      :retry_backoff_ms => @retry_backoff_ms,
      :topic_metadata_refresh_interval_ms => @topic_metadata_refresh_interval_ms,
      :queue_buffering_max_ms => @queue_buffering_max_ms,
      :queue_buffering_max_messages => @queue_buffering_max_messages,
      :queue_enqueue_timeout_ms => @queue_enqueue_timeout_ms,
      :batch_num_messages => @batch_num_messages,
      :send_buffer_bytes => @send_buffer_bytes,
      :client_id => @client_id
    }
    
    
    
    @producer = Kafka::Producer.new(options)
    @producer.connect

    @logger.info('Registering kafka producer', :topic_id => @topic_id, :broker_list => @broker_list)

    @exception_repeats = {}
    
    @codec.on_event do |event|
      begin
        @producer.send_msg(@topic_id,nil,event)
      rescue LogStash::ShutdownSignal
        @logger.info('Kafka producer got shutdown signal')
      rescue => e
        @logger.warn('kafka producer threw exception, restarting',
                     :exception => e)
        
        if @logstash_stop_on_exception_repeat > 0
            case @exception_repeats[e.to_s]
            when nil
                @exception_repeats[e.to_s] = @logstash_stop_on_exception_repeat
            when 1
                @logger.error("Exception repeated over #{@logstash_stop_on_exception_repeat} times: ",
                             :exception => e)
                finished
                raise LogStash::ShutdownSignal
            when 2..@logstash_stop_on_exception_repeat
                @exception_repeats[e.to_s] -= 1
            else
                @logger.warn('Exception repeater error')
            end 
        end
      end
    end
    
  end # def register

  def receive(event)
    return unless output?(event)
    if event == LogStash::SHUTDOWN
      finished
      return
    end
    @codec.encode(event)
  end
    
  def teardown
    @producer.close
  end

end #class LogStash::Outputs::Kafka
  
